app/bundles/LeadBundle/Entity/LeadCategoryRepository.php

```diff
@@ -55,4 +60,71 @@ public function getUnsubscribedLeadCategories(Lead $lead): array
 
         return $categories;
     }
+
+    /**
+     * @param string[] $types
+     *
+     * @return array<int, int>
+     */
+    public function getSubscribedAndNewCategoryIds(Lead $lead, array $types): array
+    {
+        $criteria = Criteria::create()
+            ->andWhere(Criteria::expr()->eq('manuallyRemoved', 1));
+
+        return $this->getLeadCategoriesMapping($lead, $types, $criteria);
+    }
+
+    /**
+     * @param string[] $types
+     *
+     * @return array<int, int>
+     */
+    public function getNonAssociatedCategoryIdsForAContact(Lead $lead, array $types): array
+    {
+        return $this->getLeadCategoriesMapping($lead, $types);
+    }
+
+    /**
+     * @param string[] $types
+     *
+     * @return array<int, int>
+     */
+    private function getLeadCategoriesMapping(Lead $lead, array $types, Criteria $criteria = null): array
+    {
+        $parentQ = $this->getEntityManager()->getRepository(Category::class)->createQueryBuilder('c');
+        $parentQ->select('c.id');
+        $parentQ->where('c.isPublished = :isPublished');
+        $parentQ->setParameter('isPublished', 1);
+        $parentQ->andWhere($parentQ->expr()->in('c.bundle', ':bundles'));
+        $parentQ->setParameter('bundles', $types, ArrayParameterType::STRING);
+
+        // Get the category ids for particular lead
+        $subQ = $this->getEntityManager()->getRepository(LeadCategory::class)->createQueryBuilder('lc');
```

Comment:
This class is already the `LeadCategoryRepository`
```suggestion
        $subQ = $this->createQueryBuilder('lc');
```

And I'd even suggest to use DBAL query builder instead of ORM query builder as in the end you need just a list of IDs. So no need to populate entities and so on.